fix: allow admin users to delete other household recipes

mariodz95 opened the pull request:
<!--
  This template provides some ideas of things to include in your PR description.

  To start, try providing a short summary of your changes in the Title above. We follow Conventional Commits syntax, please ensure your title is prefixed with one of:
  - `feat: `
  - `fix: `
  - `docs: `
  - `chore: `
  - `dev:`

  If a section of the PR template does not apply to this PR, then delete that section.

  PLEASE READ:
  -------------------------
  Mealie is moving to a regular, automatic release schedule. This means that all PRs should be in a
  stable state, ready for release. This includes:

  - Ensuring new tests have been added to cover new features, or to prevent regressions.
  - Work is fully complete and usable

 -->

## What this PR does / why we need it:

Allow admin users to delete other household recipes.

<!--
  What goal is this change working towards?
  Provide a bullet pointed summary of how each file was changed.
  Briefly explain any decisions you made with respect to the changes.
  Include anything here that you didn't include in *Release Notes*
  above, such as changes to CI or changes to internal methods.

  If there is a UI component to the change, please include before/after images.
-->

## Which issue(s) this PR fixes:

Fiexes #5437 

<!--
If this PR fixes one of more issues, list them here.
One per line, like so:
Fixes #123
Fixes #39

mariodz95 committed:
fix: allow admin users to delete other household recipes

mariodz95 committed:
Merge branch 'mealie-next' into fix/admin-users-couldnt-delete-other-household-recipes

michael-genson reviewed (changes requested):
This allows admins to _edit_ recipes belonging to other households, which I don't think we want. Similarly, just below this, we prevent admins from editing locked recipes:

```python
if recipe.settings.locked:
    return False
```

Should we instead make a separate `can_delete` method which is less restrictive to admins? Something like:
```python
if self.user.admin:
    return True
else:
    return self.can_update(recipe)
```

We should also probably add a warning to the frontend that says something like "You're about to delete a recipe that isn't yours using admin permissions. Are you sure?". I can see confusion stemming from this, especially for smaller instances where users might not even know they're admins.

mariodz95 commented:
Okay, so admins shouldn't be able to delete other household recipes if they are locked.
And if household recipes are unlocked, admin users can delete them, but non-admin users cannot? Only user who owns recipe.
Just to make sure we're on the same page.
Also, is there already some kind of modal I can follow as a pattern?

michael-genson commented:
That's the current behavior, but as long as we throw in a warning to the admin user I think it would be okay to let them delete a locked recipe (they are an admin, after all).

There's already a delete confirmation dialog (see: `RecipeActionMenu`). You would basically want to add a warning banner only if the user is an admin:

```js
<BaseDialog v-model="deleteDialog" :title="$t('recipe.delete-recipe')" color="error"
  :icon="$globals.icons.alertCircle" can-confirm @confirm="emitDelete()">
  <!-- This is the new part -->
  <BannerWarning v-if="isAdmin">
    {{ $t("recipe.some-new-string-added-to-enUS.json" }}
  </BannerWarning>
  <v-card-text>
    {{ $t("recipe.delete-confirmation") }}
  </v-card-text>
</BaseDialog>
```

Might have to tweak the style a bit.
If you haven't added a new language string before, basically just add it manually to the `en-US.json` file. We have automations that handle all other languages.

mariodz95 committed:
Merge branch 'mealie-next' into fix/admin-users-couldnt-delete-other-household-recipes

mariodz95 committed:
fix: display warning message if admin tries to delete other household recipe, remove delete option if user is not allowed to delete it

mariodz95 commented:
> That's the current behavior, but as long as we throw in a warning to the admin user I think it would be okay to let them delete a locked recipe (they are an admin, after all).
> 
> There's already a delete confirmation dialog (see: `RecipeActionMenu`). You would basically want to add a warning banner only if the user is an admin:
> 
> ```js
> <BaseDialog v-model="deleteDialog" :title="$t('recipe.delete-recipe')" color="error"
>   :icon="$globals.icons.alertCircle" can-confirm @confirm="emitDelete()">
>   <!-- This is the new part -->
>   <BannerWarning v-if="isAdmin">
>     {{ $t("recipe.some-new-string-added-to-enUS.json" }}
>   </BannerWarning>
>   <v-card-text>
>     {{ $t("recipe.delete-confirmation") }}
>   </v-card-text>
> </BaseDialog>
> ```
> 
> Might have to tweak the style a bit. If you haven't added a new language string before, basically just add it manually to the `en-US.json` file. We have automations that handle all other languages.

Added new warning message if an admin tries to delete another household recipe, and removed delete option if user is not allowed to delete it.

mariodz95 committed:
Merge branch 'mealie-next' into fix/admin-users-couldnt-delete-other-household-recipes

JPorter-02 commented:
Would it also be worth considering that the admin should be able to edit/delete locked recipes across the platform, unless the household has recipes locked? For example if a user account has been deleted but still has recipes assigned to it?

michael-genson reviewed (changes requested):
Looks good! Can we add backend tests for this? Just the basic test cases:
- admin can delete a locked recipe that isn't theirs
- admin can delete another household's recipe

And we should already have tests for these, but double check (and add them if they're missing):
- user can delete their own locked recipe
- user cannot delete someone else's locked recipe
- user can delete their own household recipe
- user cannot delete another household's recipe

If you're not sure how to check/add these let me know and I'm happy to help

michael-genson commented:
> Would it also be worth considering that the admin should be able to edit/delete locked recipes across the platform

@JPorter-02 I think this would be a good change for sure, but not necessarily for this PR

mariodz95 committed:
Merge branch 'mealie-next' into fix/admin-users-couldnt-delete-other-household-recipes

mariodz95 committed:
Merge branch 'mealie-next' into fix/admin-users-couldnt-delete-other-household-recipes

mariodz95 commented:
> Looks good! Can we add backend tests for this? Just the basic test cases:
> 
> * admin can delete a locked recipe that isn't theirs
> * admin can delete another household's recipe
> 
> And we should already have tests for these, but double check (and add them if they're missing):
> 
> * user can delete their own locked recipe
> * user cannot delete someone else's locked recipe
> * user can delete their own household recipe
> * user cannot delete another household's recipe
> 
> If you're not sure how to check/add these let me know and I'm happy to help

Just to clarify, since the test structure doesn't seem to follow the project structure, should these tests be written in test_recipe_crud.py, or is it acceptable to add a new test_recipe_service.py file? From what I can see, most of the tests you've written appear to be missing.

michael-genson commented:
For the cross-households tests, we already have tests under `test_recipe_cross_household.py`. We can add onto those.
For the non-household tests (e.g. admin can delete a locked recipe) check out `test_recipe_owner.py`. We can add onto those with admin-specific tests.

mariodz95 commented:
> For the cross-households tests, we already have tests under `test_recipe_cross_household.py`. We can add onto those. For the non-household tests (e.g. admin can delete a locked recipe) check out `test_recipe_owner.py`. We can add onto those with admin-specific tests.

Ok will check those files and try to cover all cases. That way looks much more complicated to me than if all the tests were in one file and followed the project structure. Can you give me more informations about groups and households?

michael-genson commented:
You can check out the docs regarding groups and households: https://docs.mealie.io/documentation/getting-started/features/#groups-and-households

> That way looks much more complicated to me than if all the tests were in one file and followed the project structure

Yeah our test structure is pretty messy. It's a consequence of iterative additions. Households were added pretty recently, as opposed to other core features which have been around since basically forever.

mariodz95 committed:
Merge branch 'mealie-next' into fix/admin-users-couldnt-delete-other-household-recipes

mariodz95 committed:
fix: add tests

mariodz95 committed:
Merge branch 'mealie-next' into fix/admin-users-couldnt-delete-other-household-recipes

michael-genson committed:
add admin fixture

michael-genson committed:
re-arrange tests

michael-genson committed:
Merge branch 'mealie-next' into fix/admin-users-couldnt-delete-other-household-recipes

michael-genson reviewed (approved):
LGTM - thanks for your work on this!
I moved a couple tests around to better fit the existing test structure and catch a couple edge cases. Another maintainer will have to approve this one.

Kuchenpirat reviewed (approved):
Lets get this is then 👍

klacointe commented:
What about editing of other houhsehols recipes as an admin ? Is this abandoned ?
